### ISSUE TITLE
java.lang.Character toLowerCase(codePoint: Int) / toUpperCase(codePoint: Int)

### ISSUE THREAD

**Issue by er1c** [scala-js/scala-js#4206]
Missing signatures in `java.lang.Character`:

* `def toUpperCase(codePoint: scala.Int): scala.Int`
* `def toLowerCase(codePoint: scala.Int): scala.Int`

*ECMAScript6* has all of the cool codepoint functions e.g.:

```
console.log(String.fromCodePoint(9731, 9733, 9842, 0x2F804));
// expected output: "☃★♲你"
```

But this isn't supported in *ECMAScript5*

From @sjrd (https://github.com/cquiroz/scala-java-locales/issues/197#issuecomment-697640082):

> > ```
> > * `java.lang.Character.toUpperCase(codePoint: Int)`
> > * `java.lang.Character.toLowerCase(codePoint: Int)`
> > ```
> 
> 
> Those we should be able to easily implement by delegating to the `String` equivalents, after having encoded the code points as UTF-16 code unit(s).

Another possible JVM approach: https://github.com/ekrich/scala-unicode/blob/master/src/main/scala/CaseFoldingTest.scala

## Implementation Notes

[Character.toChars(codePoint: Int): Array[Char]](https://github.com/scala-js/scala-js/blob/c8e45b21de7824230b01a4dcff978ec3464a2f06/javalanglib/src/main/scala/java/lang/Character.scala#L473) to split the codepoint & [String.new(codePoints: Array[Int], offset: Int, count: Int): String](https://github.com/scala-js/scala-js/blob/c8e45b21de7824230b01a4dcff978ec3464a2f06/javalanglib/src/main/scala/java/lang/_String.scala#L396)

[String.fromCodePoint(codePoint: Int): String](https://github.com/scala-js/scala-js/blob/c8e45b21de7824230b01a4dcff978ec3464a2f06/javalanglib/src/main/scala/java/lang/_String.scala#L449)

**Comment by ekrich** [comment]
@sjrd Is it better to delegate to JS or bring your own and match Java 8 for example? The disadvantage of matching Java is that you have the code tables and lookup code to maintain.

**Comment by er1c** [comment]
@sjrd does this approach look (generally) correct:

```
  def toUpperCase(codePoint: scala.Int): scala.Int = {
    val codePoints: Array[Char] = toChars(codePoint)
    val s = new String(codePoints, 0, codePoints.length)
    val upperCaseChars: Array[Char] = s.toUpperCase.toCharArray

    if (upperCaseChars.length == 2)
      toCodePoint(upperCaseChars(0), upperCaseChars(1))
    else
      upperCaseChars(0).toInt
  }
```

**Comment by sjrd** [comment]
> @sjrd Is it better to delegate to JS or bring your own and match Java 8 for example? The disadvantage of matching Java is that you have the code tables and lookup code to maintain.

Normally we match the JVM, and we bring our (compressed) tables. But for toLowerCase/toUpperCase we already have precedent with `String.toLowerCase()` and `String.toUpperCase()` to delegate to the JS function, so this probably what we will keep doing for those.

> @sjrd does this approach look (generally) correct:

Yes, overall that's the idea, although there are shortcuts that we can take:

* Instead of the first two lines, you can reuse [`_String.fromCodePoint`](https://github.com/scala-js/scala-js/blob/6398685a38a0f64bfde154bfdd5b8f0d54888ae9/javalanglib/src/main/scala/java/lang/_String.scala#L449-L459) (make it `private[lang]`).
* Instead of converting the string back into an `Array[Char]` with `toCharArray`, you can directly access `String.length()` and `String.charAt`, which avoids the additional allocation of an `Array[Char]` for nothing.

Then I'm not sure what we should do with results that span more than one code point. It's not necessarily just `length > 2`; it's also possible that this happens with `length == 2` if the two Chars are not a surrogate pair. We need to carefully check the spec for that.

**Comment by ekrich** [comment]
Ok, I don't think I ever considered delegating to C in Scala Native - we have parity now but at the expense of the code table memory. The upper/lower code tables in Scala Native are compressed at least to start and end of codepoint sequences. There could be better compressions and smarter lookup.

**Comment by er1c** [comment]
Added initial draft PR here: https://github.com/scala-js/scala-js/pull/4207

Still need to track down multiple-character expansion test examples

**Comment by sjrd** [comment]
Fixed in a98fcf854f1f5d0bba96432cc7bd892cc5cd06fa
